base64.go

```diff
@@ -43,6 +44,12 @@ func (b64 *Base64String) UnmarshalJSON(raw []byte) (err error) {
 	if err = json.Unmarshal(raw, &str); err != nil {
 		return
 	}
-	*b64, err = base64.RawStdEncoding.DecodeString(str)
+	// We must check whether the string was encoded in a URL-safe way in order
+	// to use the appropriate encoding.
+	if strings.ContainsAny(str, "-_") {
+		*b64, err = base64.RawURLEncoding.DecodeString(str)
```

Comment:
Won't this create problems when we encode the data? Would it make more sense to add a `URLBase64String` instead?

Reply:
> Won't this create problems when we encode the data?

I don't think so, since the characters used to detect an URL-safe base64 are specific to this encoding and are assured to not appear in a standard base64-encoded string.

> Would it make more sense to add a `URLBase64String` instead?

Not sure, it would mean we must check the encoding each time we want to decode a base64 string provided by a third party and I'm not sure it would be really relevant.

Reply:
If the spec is requiring URL-safe base64 then shouldn't we make sure that we encode using URL-safe base64 as well as decoding it?

Reply:
The spec doesn't actually mention whether the encoding has to be URL-safe (and even [uses the non-URL-safe encoding](https://matrix.org/speculator/spec/HEAD/appendices.html#unpadded-base64)), so not sure we need URL-safe encoding.
Decoding is required as sydent encodes its keys with the URL-safe encoding, but since it isn't mentioned int the spec and it doesn't change the structure of the endpoint I understand it as a liberty of the implementation.